qiskit/primitives/base_sampler.py

```diff
@@ -0,0 +1,179 @@
+
+    # parametrized circuit
+    pqc = RealAmplitudes(num_qubits=2, reps=2)
+    pqc.measure_all()
+    pqc2 = RealAmplitudes(num_qubits=2, reps=3)
+    pqc2.measure_all()
+
+    theta1 = [0, 1, 1, 2, 3, 5]
+    theta2 = [1, 2, 3, 4, 5, 6]
+    theta3 = [0, 1, 2, 3, 4, 5, 6, 7]
+
+    with Sampler(circuits=[pqc, pqc2], parameters=[pqc.parameters, pqc2.parameters]) as sampler:
+        result = sampler([0, 0, 1], [theta1, theta2, theta3])
+
+        # result of pqc(theta1)
+        print([q.binary_probabilities() for q in result[0].quasi_dists])
+
+        # result of pqc(theta2)
+        print([q.binary_probabilities() for q in result[1].quasi_dists])
+
+        # result of pqc2(theta3)
+        print([q.binary_probabilities() for q in result[2].quasi_dists])
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable, Iterable, Sequence
+
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.exceptions import QiskitError
+
+from .sampler_result import SamplerResult
+
+
+class BaseSampler(ABC):
+    """Sampler base class
+
+    Base class of Sampler that calculates quasi-probabilities of bitstrings from quantum circuits.
+    """
+
+    def __init__(
+        self,
+        circuits: Iterable[QuantumCircuit],
+        parameters: Iterable[Iterable[Parameter]] | None = None,
+    ):
+        """
+        Args:
+            circuits: quantum circuits to be executed
+            parameters: parameters of quantum circuits
+                Defaults to `[circ.parameters for circ in circuits]`
+
+        Raises:
+            QiskitError: for mismatch of circuits and parameters list.
+        """
+        self._circuits = tuple(circuits)
+        if parameters is None:
+            self._parameters = tuple(circ.parameters for circ in self._circuits)
+        else:
+            self._parameters = tuple(ParameterView(par) for par in parameters)
+            if len(self._parameters) != len(self._circuits):
+                raise QiskitError(
+                    f"Different number of parameters ({len(self._parameters)} "
+                    f"and circuits ({len(self._circuits)}"
+                )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        self.close()
+
+    @abstractmethod
+    def close(self):
+        """Close the session and free resources"""
+        ...
+
+    @property
+    def circuits(self) -> tuple[QuantumCircuit, ...]:
+        """Quantum circuits
+
+        Returns:
+            quantum circuits
+        """
+        return self._circuits
+
+    @property
+    def parameters(self) -> tuple[ParameterView, ...]:
+        """Parameters of quantum circuits
+
+        Returns:
+            Parameter list of the quantum circuits
+        """
+        return self._parameters
+
+    @abstractmethod
+    def __call__(
+        self,
+        circuits: Sequence[int],
```

Comment:
shouldn't this be optional for the sampler? by default i think it should sample all the circuits.

Reply:
It's cleaner for it to be mandatory

Reply:
I think this is open for discussion and could be made Optional. Perhaps in a subclass, I might make it Optional, but that would not violate Liskov's substitution principle. Also, I think we can allow Optional in the future without warning.